### PR TITLE
feat: add Radio and RadioGroup components

### PR DESCRIPTION
**Pull request by tinaClin**
# Radio

Radios are used when only one choice may be selected in a series of options.

Native HTML radios are 100% accessible by default, so we used a [very common CSS technique](https://dev.to/lkopacz/create-custom-keyboard-accessible-radio-buttons-22eh) to style the radio.

The Radio component composes `ControlBox`, a component we created to make it easy to style sibling inputs. [Check it out](./controlbox)

image
image
image

## Installation

```js
import Radio from '@trendmicro/react-styled-core/Radio';
// or
import { Radio } from '@trendmicro/react-styled-core';
```

### Usage

```jsx
<Radio>Radio button</Radio>
```

### Colors

Use the `variantColor` prop to change the color scheme of the Radio. `variantColor` can be any color key with key `50`(hover), `60`(checked) that exist in the them.color.

```jsx
<Stack direction="row" spacing="2x">
  <Radio variantColor="red" defaultChecked>Radio</Radio>
  <Radio variantColor="green" defaultChecked>Radio</Radio>
</Stack>
```

### Sizes

Use the `size` prop to change the size of the Radio. You can set the value to `sm`, `md`, or `lg`.

```jsx
<Stack spacing="2x">
  <Radio size="sm" name="1">Radio</Radio>
  <Radio size="md" name="1">Radio</Radio>
  <Radio size="lg" name="1">Radio</Radio>
</Stack>
```

### States

```jsx
<Stack direction="row" spacing="2x">
  <Radio>Radio</Radio>
  <Radio defaultChecked>Radio</Radio>
  <Radio disabled>Radio</Radio>
  <Radio disabled defaultChecked>Radio</Radio>
</Stack>
```

## Props

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| id | `string` | | The id assigned to input field |
| name | `string` | | The name of the input field in a radio (Useful for form submission) |
| value | `string` or `number` | | The value to be used in the radio input. This is the value that will be returned on form submission |
| variantColor | `string` | | The color of the radio when it's checked. This should be one of the color keys in the theme (e.g."green", "red") |
| size | `sm`, `md`, `lg` | `md` | The size (width and height) of the radio |
| defaultChecked | `boolean` | | If `true`, the radio will be initially checked |
| checked | `boolean` | | If `true`, the radio will be checked. You'll need to pass `onChange` to update it's value (since it's now controlled) |
| disabled | `boolean` | | If `true`, the radio will be disabled |
| children | `React.ReactNode` | | The children of the radio |
| onChange | `function` | | Function called when the state of the radio changes |
| onBlur | `function` | | Function called when you blur out of the radio |
| onFocus | `function` | | Function called when the radio receive focus |

---

# RadioGroup

image
image
image
image

## Installation

```js
import RadioGroup from '@trendmicro/react-styled-core/RadioGroup';
// or
import { RadioGroup } from '@trendmicro/react-styled-core';
```

### Usage

```jsx
function RadioExample() {
  const [value, setValue] = React.useState("2");
  return (
    <RadioGroup onChange={e => setValue(e.target.value)} value={value}>
      <Stack spacing="1x">
        <Radio value="1">First</Radio>
        <Radio value="2">Second</Radio>
        <Radio value="3">Third</Radio>
      </Stack>
    </RadioGroup>
  );
}
```

### Group orientation

Make a set of radios appear horizontal stacked rather than vertically, by adding `direction="row"` to the `Stack` component.

```jsx
<RadioGroup defaultValue="1">
  <Stack direction="row" spacing="6x">
    <Radio value="1">Radio 1</Radio>
    <Radio value="2">Radio 2</Radio>
    <Radio value="3">Radio 3</Radio>
  </Stack>
</RadioGroup>
```

### Colors

Use the `variantColor` prop to change the color scheme of the Radio. `variantColor` can be any color key with key `50`(hover), `60`(checked) that exist in the them.color.

```jsx
<RadioGroup variantColor="green" defaultValue="2">
  <Stack direction="row" spacing="6x">
    <Radio value="1">Radio 1</Radio>
    <Radio value="2">Radio 2</Radio>
  </Stack>
</RadioGroup>
```

### Sizes

Use the `size` prop to change the size of the Radio. You can set the value to `sm`, `md`, or `lg`.

```jsx
<Stack direction="column" spacing="1rem">
  <RadioGroup size="sm">
    <Stack direction="row" spacing="2x">
      <Radio value="1">Radio 1</Radio>
      <Radio value="2">Radio 2</Radio>
      <Radio value="3">Radio 3</Radio>
    </Stack>
  </RadioGroup>
  <RadioGroup size="md">
    <Stack direction="row" spacing="2x">
      <Radio value="1">Radio 1</Radio>
      <Radio value="2">Radio 2</Radio>
      <Radio value="3">Radio 3</Radio>
    </Stack>
  </RadioGroup>
  <RadioGroup size="lg">
    <Stack direction="row" spacing="2x">
      <Radio value="1">Radio 1</Radio>
      <Radio value="2">Radio 2</Radio>
      <Radio value="3">Radio 3</Radio>
    </Stack>
  </RadioGroup>
</Stack>
```

### States
```jsx
<RadioGroup defaultValue="1" disabled>
  <Stack direction="row" spacing="2x">
    <Radio value="1">Radio 1</Radio>
    <Radio value="2">Radio 2</Radio>
    <Radio value="3">Radio 3</Radio>
  </Stack>
</RadioGroup>
```

## Props

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| name | `string` | | The name used to reference the value of the control. If you don't provide this prop, it falls back to a randomly generated name. |
| value | `string` or `number` | | The value to be used in the radio input. This is the value that will be returned on form submission. |
| disabled | `boolean` | false | If `true`, all radios will be disabled. |
| variantColor | `string` | | The color of the radio when it's checked. This should be one of the color keys in the theme (e.g."green", "red"). |
| size | `sm`, `md`, `lg` | `md` | The size (width and height) of the radio. |
| defaultValue | `string` or `number` | | The default `input` element value. Use when the component is not controlled. |
| children | `React.ReactNode` | | The children of the radio. |
| onChange | `function` | | Function called when the state of the radio changes. |

### CODE REVIEW

**Comment by cheton** [comment]
Suggest using React Context instead of `cloneElement` to group Radio components.

Related to https://github.com/trendmicro-frontend/react-radio/pull/7

File: packages/react-styled-core/src/RadioGroup/index.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import Box from '../Box';
+
+const RadioGroup = forwardRef(
+  (
+    {
+      onChange,
+      name,
+      value: valueProp,
+      defaultValue,
+
```
Comment (austinhung81) [inline packages/react-styled-core/src/RadioGroup/index.js]: I'm just curious why there has a new line, lolll

File: packages/react-styled-core/src/RadioGroup/index.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      if (!isValidElement(child)) {
+        return null;
+      }
+
+      const isLastRadio = (children.length === index + 1);
+      const defaultSpacing = spacing || (inline ? '6x' : '1x');
+      const spacingProps = inline ? { mr: defaultSpacing } : { mb: defaultSpacing };
+
+      return (
+        <Box
```
Comment (cheton) [inline packages/react-styled-core/src/RadioGroup/index.js]: Box is not required here if we wrap children with a context provider.

File: packages/react-styled-core/src/RadioGroup/index.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+          display={inline ? 'inline-block' : 'block'}
```
Comment (cheton) [inline packages/react-styled-core/src/RadioGroup/index.js]: If so, you don't need to pass the `inline` prop to align radio components.

File: packages/react-styled-core/src/RadioGroup/index.js (hunk 1 above)
Comment (tinaClin) [reply packages/react-styled-core/src/RadioGroup/index.js]: 只是覺得這樣子分開比較好看...XD

File: packages/react-styled-core/src/RadioGroup/index.js (hunk 2 above)
Comment (tinaClin) [reply packages/react-styled-core/src/RadioGroup/index.js]: Add `GroupContext/index.js` file to implement this way.

Comment (tinaClin) [reply packages/react-styled-core/src/RadioGroup/index.js]: 這個部分，我會另外發一個對CheckboxGroup的PR

File: packages/react-styled-core/src/RadioGroup/index.js (hunk 3 above)
Comment (tinaClin) [reply packages/react-styled-core/src/RadioGroup/index.js]: Done

**Review by cheton — APPROVED** [review]

File: packages/react-styled-core/src/RadioGroup/index.js (hunk 2 above)
Comment (cheton) [reply packages/react-styled-core/src/RadioGroup/index.js]: ok

**Pull request merged**